internal/scheduler/scheduler.go

```diff
@@ -2,344 +2,657 @@ package scheduler
 	return &Scheduler{
-		Producer:              producer,
-		Db:                    db,
-		ActiveJobs:            make(map[uuid.UUID]*schedulerdb.Job),
-		RunsByJobId:           make(map[uuid.UUID]*JobRuns),
-		Nodes:                 make(map[string]*schedulerdb.Nodeinfo),
-		Executors:             make(map[string]time.Time),
-		ExecutorAliveDuration: 5 * time.Minute,
-	}
+		jobRepository:      jobRepository,
+		executorRepository: executorRepository,
+		schedulingAlgo:     schedulingAlgo,
+		leaderController:   leaderController,
+		publisher:          publisher,
+		jobDb:              jobDb,
+		clock:              clock.RealClock{},
+		cyclePeriod:        cyclePeriod,
+		executorTimeout:    executorTimeout,
+		maxLeaseReturns:    maxLeaseReturns,
+		jobsSerial:         -1,
+		runsSerial:         -1,
+	}, nil
 }
 
-func (srv *Scheduler) Run(ctx context.Context) error {
-	// Get the configured logger, or the standard logger if none is provided.
-	var log *logrus.Entry
-	if srv.Logger != nil {
-		log = srv.Logger.WithField("service", "SchedulerIngester")
-	} else {
-		log = logrus.StandardLogger().WithField("service", "SchedulerIngester")
+// Run enters the scheduling loop, which will continue until the ctx is cancelled
+func (s *Scheduler) Run(ctx context.Context) error {
+	//  Do initial population of Job Db
+	err := s.initialise(ctx)
+	if err != nil {
+		return err
 	}
-	log.Info("service started")
 
-	refreshTicker := time.NewTicker(10 * time.Second)
-	scheduleTicker := time.NewTicker(11 * time.Second)
+	ticker := s.clock.NewTicker(s.cyclePeriod)
+	log.Infof("Will run scheduling cycle every %s", s.cyclePeriod)
+	prevLeaderToken := LeaderToken{leader: false, id: uuid.New()}
 	for {
 		select {
 		case <-ctx.Done():
-			return ctx.Err()
-		case <-refreshTicker.C:
-			g, groupCtx := errgroup.WithContext(ctx)
-			entry := log.WithFields(logrus.Fields{
-				"JobsSerial":  srv.JobsSerial,
-				"RunsSerial":  srv.RunsSerial,
-				"NodesSerial": srv.NodesSerial,
-			})
-			ctxWithLogger := ctxlogrus.ToContext(groupCtx, entry)
-			g.Go(func() error { return srv.updateJobsRuns(ctxWithLogger) })
-			g.Go(func() error { return srv.updateNodes(ctxWithLogger) })
-			err := g.Wait()
-			if err != nil {
-				logging.WithStacktrace(log, err).Info("failed to read from postgres")
+			log.Infof("Context cancelled, returning..")
+			return nil
+		case <-ticker.C():
+			start := s.clock.Now()
+			leaderToken := s.leaderController.GetToken()
+			fullUpdate := false
+			log.Infof("Received a leaderToken. Leader status is %t", leaderToken.leader)
+
+			// If we are becoming leader then we must ensure we have caught up to all Pulsar messages
+			if leaderToken.leader && leaderToken != prevLeaderToken {
+				log.Infof("Becoming leader")
+				fullUpdate = true
+				err := s.ensureDbUpToDate(ctx, 1*time.Second)
+				if err != nil {
+					log.WithError(err).Error("Could not become master")
+					leaderToken = InvalidLeaderToken()
```

Comment:
How is the k8s leader election logic notified that we're no longer leader so it can trigger a failover?

Reply:
It's not.  As far as K8's is concerned you are still leader.  Net time round the scheduling cycle you'll be told you're leader again and you can have another go at syncing you db state.

We could try and notify  K8s that your not leader, but I haven't implemented that yet.  A poor man's version of that would be to just crash at this point.

Reply:
Maybe we should blow up/return with an error here then?

Reply:
> Maybe we should blow up/return with an error here then?
Yeah we could- the disadvantage is that we'd put more load on the db, because K8s would just restart us and then we'd hit the db again to re-initialise from scratch.

I think the current logic is fine- but We can re-evaluate this once I've done some E2E testing with real k8s.